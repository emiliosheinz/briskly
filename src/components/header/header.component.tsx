import { signIn, signOut, useSession } from 'next-auth/react'
import Image from 'next/image'
import Link from 'next/link'

// TODO emiliosheinz: Create unit tests for Header component
// TODO emiliosheinz: Create Storybook documentation for Header component
export function Header() {
  const { data: sessionData } = useSession()

  const onClick = sessionData ? () => signOut() : () => signIn()
  const label = sessionData ? 'Sign out' : 'Sign in'

  function renderAuthButton() {
    return (
      <button
        className='hover:text-gray-900 whitespace-nowrap text-base font-medium text-primary-900'
        onClick={onClick}
      >
        {label}
      </button>
    )
  }

  return (
    <header className='relative bg-primary-50'>
      <div className='mx-auto max-w-7xl px-4 sm:px-6'>
        <div className='flex items-center justify-between border-b-2 border-primary-200 py-6 md:justify-start md:space-x-10'>
          <div className='flex justify-start lg:w-0 lg:flex-1'>
            <Link href='/'>
              <span className='sr-only'>Briskly</span>
              {/* TODO emiliosheinz: Change image component to use custom component */}
              <Image
                className='h-8 w-auto sm:h-10'
                src='/images/logo.png'
                width={80}
                height={80}
                alt='Briskly logo'
              />
            </Link>
          </div>
          <div className='flex flex-1 items-center justify-end lg:w-0'>
            {renderAuthButton()}
          </div>
        </div>
      </div>
    </header>
  )
}
