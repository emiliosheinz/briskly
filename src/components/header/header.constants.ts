import {
  ClockIcon,
  StarIcon,
  FireIcon,
  HeartIcon,
  PlusCircleIcon,
  UserCircleIcon,
  WrenchIcon,
} from '@heroicons/react/24/outline'

import { routes } from '~/utils/navigation'

const MENU_OPTIONS = [
  {
    label: 'Para Revisar',
    icon: ClockIcon,
    href: routes.toBeReviewed(),
    isAuthRequired: true,
  },
  {
    label: 'Mais Votados',
    icon: StarIcon,
    href: '#',
    isAuthRequired: false,
  },
  {
    label: 'Mais Recentes',
    icon: FireIcon,
    href: '#',
    isAuthRequired: false,
  },
  {
    label: 'Para Você',
    icon: HeartIcon,
    href: '#',
    isAuthRequired: true,
  },
  {
    label: 'Criar Deck',
    icon: PlusCircleIcon,
    href: routes.createNewDeck(),
    isAuthRequired: false,
  },
  {
    label: 'Meu Perfil',
    icon: UserCircleIcon,
    href: '#',
    isAuthRequired: true,
  },
  {
    label: 'Configurações',
    icon: WrenchIcon,
    href: '#',
    isAuthRequired: true,
  },
]

export const MOBILE_MENU_OPTIONS = [...MENU_OPTIONS]

export const DESKTOP_MENU_OPTIONS = {
  PRIMARY: MENU_OPTIONS.filter(({ isAuthRequired }) => !isAuthRequired),
  SECONDARY: MENU_OPTIONS.filter(({ isAuthRequired }) => isAuthRequired),
}
