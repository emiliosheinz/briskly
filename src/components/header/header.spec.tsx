import { render, screen } from '~/utils/tests/react-testing-library'
import { withNextAuth } from '~/utils/tests/with-next-auth'

import { Header } from './header.component'

describe('Header Component', () => {
  describe('Loading Authentication', () => {
    it('should be correctly rendered', () => {
      const component = render(withNextAuth(<Header />, 'loading'))

      expect(component.asFragment()).toMatchSnapshot()
    })

    it('should render a link to the website home page', () => {
      render(withNextAuth(<Header />, 'loading'))

      const homeLink = screen.getByTestId('home-anchor')

      expect(homeLink).toBeInTheDocument()
    })

    it('should render only DesktopPrimaryMenu and MobileHamburgerMenu', () => {
      render(withNextAuth(<Header />, 'loading'))

      const desktopPrimaryMenu = screen.getByTestId('desktop-primary-menu')
      const mobileHamburgerMenu = screen.getByTestId('mobile-hamburger-menu')
      const desktopSecondaryMenu = screen.queryByTestId(
        'desktop-secondary-menu',
      )
      const authButton = screen.queryByTestId('auth-button')

      expect(desktopPrimaryMenu).toBeInTheDocument()
      expect(mobileHamburgerMenu).toBeInTheDocument()
      expect(desktopSecondaryMenu).not.toBeInTheDocument()
      expect(authButton).not.toBeInTheDocument()
    })
  })
  describe('Authenticated', () => {
    it('should be correctly rendered', () => {
      const component = render(withNextAuth(<Header />, 'authenticated'))

      expect(component.asFragment()).toMatchSnapshot()
    })

    it('should render a link to the website home page', () => {
      render(withNextAuth(<Header />, 'authenticated'))

      const homeLink = screen.getByTestId('home-anchor')

      expect(homeLink).toBeInTheDocument()
    })

    it('should render all the menus and no AuthButton', () => {
      render(withNextAuth(<Header />, 'authenticated'))

      const desktopPrimaryMenu = screen.getByTestId('desktop-primary-menu')
      const mobileHamburgerMenu = screen.getByTestId('mobile-hamburger-menu')
      const desktopSecondaryMenu = screen.getByTestId('desktop-secondary-menu')
      const authButton = screen.queryByTestId('auth-button')

      expect(desktopPrimaryMenu).toBeInTheDocument()
      expect(mobileHamburgerMenu).toBeInTheDocument()
      expect(desktopSecondaryMenu).toBeInTheDocument()
      expect(authButton).not.toBeInTheDocument()
    })
  })
  describe('Unauthenticated', () => {
    it('should be correctly rendered', () => {
      const component = render(withNextAuth(<Header />, 'unauthenticated'))

      expect(component.asFragment()).toMatchSnapshot()
    })

    it('should render a link to the website home page', () => {
      render(withNextAuth(<Header />, 'unauthenticated'))

      const homeLink = screen.getByTestId('home-anchor')

      expect(homeLink).toBeInTheDocument()
    })

    it('should render DesktopPrimaryMenu, MobileHamburgerMenu and AuthButton', () => {
      render(withNextAuth(<Header />, 'unauthenticated'))

      const desktopPrimaryMenu = screen.getByTestId('desktop-primary-menu')
      const mobileHamburgerMenu = screen.getByTestId('mobile-hamburger-menu')
      const desktopSecondaryMenu = screen.queryByTestId(
        'desktop-secondary-menu',
      )
      const authButton = screen.getByTestId('auth-button')

      expect(desktopPrimaryMenu).toBeInTheDocument()
      expect(mobileHamburgerMenu).toBeInTheDocument()
      expect(desktopSecondaryMenu).not.toBeInTheDocument()
      expect(authButton).toBeInTheDocument()
    })
  })
})
