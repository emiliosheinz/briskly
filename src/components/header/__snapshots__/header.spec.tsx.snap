// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component Authenticated should be correctly rendered 1`] = `
<DocumentFragment>
  <header
    class="border-b-2 border-primary-200 bg-primary-50"
  >
    <div
      class="mx-auto flex h-20 max-w-7xl items-center justify-between px-4 md:space-x-10"
    >
      <div
        class="flex items-center justify-start"
      >
        <a
          data-testid="home-anchor"
          href="/"
        >
          <span
            class="sr-only"
          >
            Briskly
          </span>
          <img
            alt="Briskly logo"
            data-nimg="1"
            data-testid="image-component"
            decoding="async"
            height="35"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Flogo.png&w=96&q=75"
            srcset="/_next/image?url=%2Fimages%2Flogo.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Flogo.png&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
        </a>
        <div
          class="mx-5 hidden gap-6 md:flex lg:mx-10 lg:gap-8"
          data-testid="desktop-primary-menu"
        >
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Mais Votados
          </a>
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Mais Recentes
          </a>
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Criar Deck
          </a>
        </div>
      </div>
      <div
        class="md:hidden"
        data-headlessui-state=""
        data-testid="mobile-hamburger-menu"
      >
        <button
          aria-expanded="false"
          class="inline-flex items-center justify-center bg-primary-50 p-1 text-primary-900"
          data-headlessui-state=""
          data-testid="popover-button"
          id="headlessui-popover-button-:r6:"
          type="button"
        >
          <span
            class="sr-only"
          >
            Open menu
          </span>
          <svg
            aria-hidden="true"
            class="h-8 w-8"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        class="relative hidden md:block"
        data-headlessui-state=""
        data-testid="desktop-secondary-menu"
      >
        <button
          aria-expanded="false"
          class="inline-flex items-center justify-center bg-primary-50 p-1 text-primary-900"
          data-headlessui-state=""
          data-testid="popover-button"
          id="headlessui-popover-button-:r8:"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="h-8 w-8 text-primary-900"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M17.982 18.725A7.488 7.488 0 0012 15.75a7.488 7.488 0 00-5.982 2.975m11.963 0a9 9 0 10-11.963 0m11.963 0A8.966 8.966 0 0112 21a8.966 8.966 0 01-5.982-2.275M15 9.75a3 3 0 11-6 0 3 3 0 016 0z"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Header Component Loading Authentication should be correctly rendered 1`] = `
<DocumentFragment>
  <header
    class="border-b-2 border-primary-200 bg-primary-50"
  >
    <div
      class="mx-auto flex h-20 max-w-7xl items-center justify-between px-4 md:space-x-10"
    >
      <div
        class="flex items-center justify-start"
      >
        <a
          data-testid="home-anchor"
          href="/"
        >
          <span
            class="sr-only"
          >
            Briskly
          </span>
          <img
            alt="Briskly logo"
            data-nimg="1"
            data-testid="image-component"
            decoding="async"
            height="35"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Flogo.png&w=96&q=75"
            srcset="/_next/image?url=%2Fimages%2Flogo.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Flogo.png&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
        </a>
        <div
          class="mx-5 hidden gap-6 md:flex lg:mx-10 lg:gap-8"
          data-testid="desktop-primary-menu"
        >
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Mais Votados
          </a>
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Mais Recentes
          </a>
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Criar Deck
          </a>
        </div>
      </div>
      <div
        class="md:hidden"
        data-headlessui-state=""
        data-testid="mobile-hamburger-menu"
      >
        <button
          aria-expanded="false"
          class="inline-flex items-center justify-center bg-primary-50 p-1 text-primary-900"
          data-headlessui-state=""
          data-testid="popover-button"
          id="headlessui-popover-button-:r0:"
          type="button"
        >
          <span
            class="sr-only"
          >
            Open menu
          </span>
          <svg
            aria-hidden="true"
            class="h-8 w-8"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
    </div>
  </header>
</DocumentFragment>
`;

exports[`Header Component Unauthenticated should be correctly rendered 1`] = `
<DocumentFragment>
  <header
    class="border-b-2 border-primary-200 bg-primary-50"
  >
    <div
      class="mx-auto flex h-20 max-w-7xl items-center justify-between px-4 md:space-x-10"
    >
      <div
        class="flex items-center justify-start"
      >
        <a
          data-testid="home-anchor"
          href="/"
        >
          <span
            class="sr-only"
          >
            Briskly
          </span>
          <img
            alt="Briskly logo"
            data-nimg="1"
            data-testid="image-component"
            decoding="async"
            height="35"
            loading="lazy"
            src="/_next/image?url=%2Fimages%2Flogo.png&w=96&q=75"
            srcset="/_next/image?url=%2Fimages%2Flogo.png&w=48&q=75 1x, /_next/image?url=%2Fimages%2Flogo.png&w=96&q=75 2x"
            style="color: transparent;"
            width="40"
          />
        </a>
        <div
          class="mx-5 hidden gap-6 md:flex lg:mx-10 lg:gap-8"
          data-testid="desktop-primary-menu"
        >
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Mais Votados
          </a>
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Mais Recentes
          </a>
          <a
            class="flex items-center space-x-1 whitespace-nowrap text-primary-900"
            href="#"
          >
            Criar Deck
          </a>
        </div>
      </div>
      <div
        class="md:hidden"
        data-headlessui-state=""
        data-testid="mobile-hamburger-menu"
      >
        <button
          aria-expanded="false"
          class="inline-flex items-center justify-center bg-primary-50 p-1 text-primary-900"
          data-headlessui-state=""
          data-testid="popover-button"
          id="headlessui-popover-button-:ri:"
          type="button"
        >
          <span
            class="sr-only"
          >
            Open menu
          </span>
          <svg
            aria-hidden="true"
            class="h-8 w-8"
            fill="none"
            stroke="currentColor"
            stroke-width="1.5"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.75 6.75h16.5M3.75 12h16.5m-16.5 5.25h16.5"
              stroke-linecap="round"
              stroke-linejoin="round"
            />
          </svg>
        </button>
      </div>
      <div
        class="hidden w-32 md:block"
      >
        <div
          class="relative"
        >
          <button
            class="flex flex-row gap-2 rounded-md px-6 py-2 text-base font-normal capitalize tracking-wide ring-2 transition-colors duration-150 ease-in-out hover:shadow-md disabled:opacity-50 bg-primary-50 enabled:hover:bg-primary-100 text-primary-900 ring-primary-900 w-full justify-center"
            data-testid="auth-button"
          >
            Login
          </button>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
